accepted/2024/trusted-publishers-oidc-for-nuget-push.md

```diff
@@ -0,0 +1,568 @@
+
+In short, a GitHub Actions trust policy is restricted in two ways:
+
+1. **Which GitHub Actions repository and workflow is trusted?**
+2. **What NuGet API operations can be used inside the GitHub actions workflow?**
+
+#### Which GitHub Actions repository and workflow is trusted?
+
+The user will sign in to NuGet.org, navigate to a new "Trusted Publishers" account page, and configure a trusted GitHub
+repository using a form like this:
+
+![Trust policy form](meta/trusted-publishers-oidc-for-nuget-push/trust-policy-form.png)
+
+In the form, the package author would provide 3 critical pieces of information:
+
+1. The GitHub organization or username owning the repository.
+2. The GitHub repository name.
+3. Some filter on which GitHub Action workflow runs should be allowed to push packages.
+
+The filter on which workflow runs should be allowed to publish packages can potentially be very specific. GitHub Actions
+supplies a variety of metadata in the OIDC token available to workflow runs. Refer to [Understanding the OIDC
+token](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)
+to see all of the properties.
+
+For the sake of simplicity, we will start with the following workflow filters:
+- Filtering for a specific *branch*, as shown in the UI above (branch filter)
+- Filtering for a specific *environment* (a property optionally set in the workflow)
+- Filtering for a specific *workflow path* (relative file path within the repository)
+
+The screenshot above only shows the branch filter, so the UI will need to be expanded to cover the desired workflow
+filters. A suggest UI would have 3 checkboxes:
+- [ ] Filter by branch: __________
+- [ ] Filter by environment: __________
+- [ ] Filter by workflow: __________
+
+At least one filter would need to be checked as to prevent an overly broad trust policy.
+
+Additional filters can be added in the future based on other properties provided in the GitHub OIDC token.
+
+In short, by adding this trust policy, the package author is saying:
+
+> I trust GitHub to only issue tokens for my repository for my GitHub Actions workflow *and* these tokens are sufficient
+> credentials for uploading packages to my NuGet.org account.
+
+Once the package author has added this trust policy, NuGet.org will trust GitHub Actions OIDC tokens matching this
+pattern much like an existing long-lived API key, with the added benefit of the trust policy never expiring.
+
+NuGet package sources must ensure a secure, authenticated session for the user that is adding the trust policy.
+NuGet.org requires two-factor authentication for all web UI sign-ins.
+
+#### What NuGet API operations can be used inside the GitHub Actions workflow?
+
+To scope the ways in which a GitHub trust policy can be used, similar scoping rules to the current API key flow can be
```

Comment:
I recommend investigating the current usage of API Key scopes to understand how they are used. For example, are globs commonly used? Are there API keys that only allow pushing new versions? Are they common? I have a suspicion that this functionality isn't used, so we should check before carrying it forward. We might want to make changes / simplify the MVP.

Reply:
Great point. I was assuming parity would be the easiest since we would be stamping out short-lived API keys from the OIDC token, so all of the existing API key scoping can apply. But there is still cost to adding these parameters to the trust policy definition and making the UI work and look pretty+accessible. I will check the usage and modify the spec if adoption is "low".

Reply:
I looked at the data. I think we can definitely get by without package ID and allowed action scoping, especially since the API keys are short-lived.